src/lib/raft/raft.c

```diff
@@ -695,14 +725,45 @@ raft_worker_handle_io(struct raft *raft)
 				raft_sm_wait_election_end(raft);
 			} else {
 				/* No leaders, no votes. */
-				raft_sm_schedule_new_vote(raft, raft->self);
+				raft_sm_schedule_new_vote(raft, raft->self,
+							  raft->vclock);
 			}
 		}
 	} else {
 		memset(&req, 0, sizeof(req));
 		assert(raft->volatile_term >= raft->term);
-		req.term = raft->volatile_term;
+		if (raft->volatile_vote == 0)
+			goto do_dump;
+		/*
+		 * Vote and term bumps are persisted separately. This serves as
+		 * a flush of all transactions going to WAL right now so as the
+		 * current node could correctly compare its own vclock vs
+		 * candidate's one. Otherwise the local vclock can be <=
+		 * candidate's now, but that can change after the WAL queue is
+		 * flushed.
+		 */
+		if (raft->volatile_term > raft->term)
+			goto do_dump;
+		/*
+		 * Skip self. When vote was issued, own vclock could be smaller,
+		 * but that doesn't matter. Can always vote for self. Not having
+		 * this special case still works, but the node might log that
+		 * it canceled a vote for self, which is confusing.
```

Comment:
> Not having this special case still works, but the node might log that it canceled a vote for self, which is confusing.

I'm not sure how it's supposed to work without that special case. Seems `raft.vote` is zero, and `raft->votalile_vote` is `self`. This shouldn't make the node become candidate. Besides, the node might forget that it voted for self after a restart.

Let's just drop that comment part.

Reply:
> This shouldn't make the node become candidate

The node is a candidate if it is configured as a candidate. In `end_dump:` it will see that vote is zero, no leader, the node is a candidate, and it simply retries voting for self. I discovered it when tried to drop the special case to see what happens.

I would rather keep the comment, otherwise in the future we won't remember all the reasons why this check is needed. I tried to reword it:
```Diff
@@ -747,8 +747,9 @@ end_dump:
 		/*
 		 * Skip self. When vote was issued, own vclock could be smaller,
 		 * but that doesn't matter. Can always vote for self. Not having
-		 * this special case still works, but the node might log that
-		 * it canceled a vote for self, which is confusing.
+		 * this special case still works if the node is configured as a
+		 * candidate, but the node might log that it canceled a vote for
+		 * self, which is confusing.
 		 */
 		if (raft->volatile_vote == raft->self)
 			goto do_dump_with_vote;
```

Reply:
> The node is a candidate if it is configured as a candidate. In end_dump: it will see that vote is zero, no leader, the node is a candidate, and it simply retries voting for self. I discovered it when tried to drop the special case to see what happens.

I see now, thanks for the explanation!